### PR TITLE
Added VipGroups (11.10)

### PR DESCRIPTION
**Pull request by reyaleman**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Added vipgroups functionality. (v11.10)

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: src/configmanager.cpp [hunk 1]
```diff
@@ -295,6 +295,8 @@ bool ConfigManager::load()
 	integer[DEPOT_PREMIUM_LIMIT] = getGlobalNumber(L, "depotPremiumLimit", 15000);
 	integer[QUEST_TRACKER_FREE_LIMIT] = getGlobalNumber(L, "questTrackerFreeLimit", 10);
 	integer[QUEST_TRACKER_PREMIUM_LIMIT] = getGlobalNumber(L, "questTrackerPremiumLimit", 15);
+	integer[VIPGROUP_FREE_LIMIT] = getGlobalNumber(L, "vipGroupFreeLimit", 3);
```
Comment (EPuncker) [inline src/configmanager.cpp:298]: use same default values as config.lua file

File: src/luascript.cpp [hunk 2]
```diff
@@ -11165,6 +11166,19 @@ int LuaScriptInterface::luaGroupGetMaxVipEntries(lua_State* L)
...
 }
 
+int LuaScriptInterface::luaGroupGetMaxVipGroups(lua_State* L)
+{
+	// group:getMaxVipGroups()
+	Group* group = getUserdata<Group>(L, 1);
+	if (group) {
+		lua_pushnumber(L, group->maxVipGroups);
+	}
+	else {
```
Comment (EPuncker) [inline src/luascript.cpp:11176]:
```suggestion
	} else {
```

File: src/luascript.cpp [hunk 3]
```diff
@@ -11165,6 +11166,19 @@ int LuaScriptInterface::luaGroupGetMaxVipEntries(lua_State* L)
 	return 1;
 }
 
+int LuaScriptInterface::luaGroupGetMaxVipGroups(lua_State* L)
+{
+	// group:getMaxVipGroups()
+	Group* group = getUserdata<Group>(L, 1);
+	if (group) {
+		lua_pushnumber(L, group->maxVipGroups);
+	}
```
Comment (EPuncker) [inline src/luascript.cpp:11175]:
```suggestion
```

**Review by EPuncker — CHANGES_REQUESTED** [review]
and please add the schema changes to schema.sql too

File: data/migrations/32.lua [hunk 4]
```diff
@@ -1,3 +1,33 @@
 function onUpdateDatabase()
-	return false
+	print("> Updating database to version 32 (vipgroups)")
+	db.query("CREATE TABLE IF NOT EXISTS `account_vipgroups` (`id` int NOT NULL AUTO_INCREMENT, `account_id` int NOT NULL, `name` varchar(128) NOT NULL DEFAULT '', `editable` tinyint NOT NULL DEFAULT '1', PRIMARY KEY (`id`), FOREIGN KEY (`account_id`) REFERENCES `accounts` (`id`) ON DELETE CASCADE ON UPDATE CASCADE) ENGINE=InnoDB;")
+	db.query("ALTER TABLE `account_viplist` ADD `id` int NOT NULL AUTO_INCREMENT PRIMARY KEY FIRST")
+	db.query("ALTER TABLE `account_viplist` RENAME `account_vipentries`")
```
Comment (DSpeichert) [inline data/migrations/32.lua:5]: I'd leave that as-is for backwards compatibility.

File: data/migrations/32.lua [hunk 5]
```diff
@@ -1,3 +1,33 @@
...
+			db.query(stmt)
+		end
+	end	
+
+	print("Run this query in your database to create the oncreate_accounts trigger:")
+	print("DELIMITER //")
+	print("CREATE TRIGGER `oncreate_accounts` AFTER INSERT ON `accounts`")
+	print(" FOR EACH ROW BEGIN")
+	print("  INSERT INTO `account_vipgroups` (`account_id`, `name`, `editable`) VALUES (NEW.id, 'Enemies', 0), (NEW.id, 'Friends', 0), (NEW.id, 'Trading Partners', 0);")
+	print("END //")
```
Comment (DSpeichert) [inline data/migrations/32.lua:30]: Why not run that in migration?

File: schema.sql [hunk 6]
```diff
@@ -362,10 +373,19 @@ CREATE TABLE IF NOT EXISTS `towns` (
   UNIQUE KEY `name` (`name`)
 ) ENGINE=InnoDB DEFAULT CHARACTER SET=utf8;
 
-INSERT INTO `server_config` (`config`, `value`) VALUES ('db_version', '31'), ('motd_hash', ''), ('motd_num', '0'), ('players_record', '0');
+CREATE TABLE IF NOT EXISTS `vipgroup_vipentry` (
```
Comment (DSpeichert) [inline schema.sql:376]: `account_vipgroup_entry` is a better name, it would list lexically right next to `account_vipgroups`

**Review by DSpeichert — CHANGES_REQUESTED** [review]

File: src/iologindata.cpp [hunk 7]
```diff
@@ -990,19 +1011,38 @@ bool IOLoginData::hasBiddedOnHouse(uint32_t guid)
 	return db.storeQuery(fmt::format("SELECT `id` FROM `houses` WHERE `highest_bidder` = {:d} LIMIT 1", guid)).get();
 }
 
+std::forward_list<VIPGroup> IOLoginData::getVIPGroups(uint32_t accountId)
```
Comment (ranisalt) [inline src/iologindata.cpp:1014]: Use a vector here, no need to use a forward list which wastes memory

**Review by ranisalt — CHANGES_REQUESTED** [review]

**Review by LukSrT — CHANGES_REQUESTED** [review]

**Comment by ranisalt** [comment]
Is someone willing to revive this PR? Apart from the comments, it should also use const& vectors to avoid copies, and needs to be reformated.

**Comment by reyaleman** [comment]
> Is someone willing to revive this PR? Apart from the comments, it should also use const& vectors to avoid copies, and needs to be reformated.

It's ok if I close this pull request and open a new one with the changes requested? I tried to rebase this but failed 😂

**Comment by ranisalt** [comment]
> It's ok if I close this pull request and open a new one with the changes requested? I tried to rebase this but failed 😂

Go ahead!

**Pull request closed without merging**